pkg/ring/model.go

```diff
@@ -45,18 +45,17 @@ func (d *Desc) AddIngester(id, addr string, tokens []uint32, state IngesterState
 		Addr:      addr,
 		Timestamp: time.Now().Unix(),
 		State:     state,
+		Tokens:    tokens,
 	}
 
-	if normaliseTokens {
-		ingester.Tokens = tokens
-	} else {
-		for _, token := range tokens {
-			d.Tokens = append(d.Tokens, TokenDesc{
-				Token:    token,
-				Ingester: id,
-			})
+	// delete any reference from d.Tokens (can happen if denormalized tokens for this ingester
```

Comment:
I'm missing some context. Can you help me understand this change, so that I can make a more conscious review, please?

Reply:
See pkg/ring/lifecycler.go:549, autoJoin method.

When ingester finds that it already owns some tokens during join, it will use those tokens, instead of generating new ones. Those tokens could have been normalized or not. If they were not normalized (i.e. stored in d.Tokens), we need to remove them, otherwise they would be stored at two places (denormalized in `d.Tokens`, but also normalized tokens in `Ingesters[id].Tokens`).

Reply:
Also, this behavior is used by updated test.

Reply:
Right, got it, thanks!